### ISSUE TITLE
Running the exact docker file and sequence of commands results in bad image

### ISSUE THREAD

**Issue by ablaylock** [MicrosoftDocs/visualstudio-docs#3281]
When I run the exact sequence of commands and docker file described here the docker build succeeds. When I go to use the container I get an error "The system cannot find the path specified." when trying to run the container. I then added the collect.exe step from the advanced example and have it run after the install regardless to collect the installation logs. When I review the logs I get a message "VS setup process exited with code 255" in the dd_bootstrapper_*.log. This suggests that the docker script does not correctly get the error from the installer.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: b91a4a43-5353-97cf-7532-7644aa3b497f
* Version Independent ID: 858d9a5c-72f3-6a85-9270-0a0dd9e0438b
* Content: [Install Visual Studio Build Tools into a container](https://docs.microsoft.com/en-us/visualstudio/install/build-tools-container?view=vs-2019#feedback)
* Content Source: [docs/install/build-tools-container.md](https://github.com/MicrosoftDocs/visualstudio-docs/blob/master/docs/install/build-tools-container.md)
* Product: **visual-studio-windows**
* Technology: **vs-installation**
* GitHub Login: @heaths
* Microsoft Alias: **tglee**

**Comment by heaths** [comment]
Exit code 255 is typically associated with either running a container image that is not compatible with your host, or in a container in which you tried to install .NET Framework which is not supported (almost always requires a reboot, which containers do not support). Unfortunately, we are not able to provide samples for the myriad of different Windows host version and container versions required, which also vary depending on whether you're using Hyper-V isolation or process isolation (the latter is available only on Windows Server at this time).

**Comment by ablaylock** [comment]
Is it possible to get some guidance for selecting the proper container version for an image? What criteria need to be considered?

**Comment by heaths** [comment]
For process isolation, the host and container image version of Windows must match exactly. For Hyper-V isolation, the same version or 1 version older is supported. See https://docs.microsoft.com/virtualization/windowscontainers/deploy-containers/version-compatibility. And you must use the appropriately tagged `dotnet-framework` image for that version of Windows, which are now on mcr.microsoft.com, though I believe they are still advertised on hub.docker.com (just not available there for newer Windows versions). See https://devblogs.microsoft.com/dotnet/net-core-container-images-now-published-to-microsoft-container-registry/ for details.

**Comment by heaths** [comment]
I'll update the doc with one or both of these links as a note.

**Comment by ablaylock** [comment]
Just for clarity in my own mind can you confirm that I have correctly interpreted the table. I am running Windows 10 Version 1809 Build 17763.504. This means I can use images based on 
* Windows Server 2016 Builds: 14393.*
* Windows Server, version 1709 Builds 16299.*
* Windows Server, version 1803 Builds 17134.*
* Windows Server 2019Builds 17763.*

Is that correct?

**Comment by ablaylock** [comment]
Additionally which of the `dotnet` images are appropriate? Which of the following are valid base images:
* .NET Core SDK
* ASP.NET Core runtime
* .NET Core Runtime
* .NET Core Runtime dependencies

**Comment by heaths** [comment]
With Hyper-V isolation, the host/container matrix should be correct. But as I mentioned, you need `dotnet-framework` images, now on MCR named `dotnet/framework`. The will have tags that denote which container OS versions they are based on, like those listed down the page at https://hub.docker.com/_/microsoft-dotnet-framework-sdk/. You'll need images with the .NET **Framework** SDK (Framework and Core are not the same; Core can be installed in a container unlike Framework). For example, `mcr.microsoft.com/dotnet/framework/sdk:4.8-windowsservercore-1803` is the latest version for Windows 10 / Server 2019 build 1803.

**Comment by ablaylock** [comment]
Ok, so the following base images should work for my host with hyperv isolation given the host/container matrix:

* mcr.microsoft.com/dotnet/framework/sdk:4.8-20190514-windowsservercore-ltsc2019 (based on 10.0.17763.504)
* mcr.microsoft.com/dotnet/framework/sdk:4.8-20190514-windowsservercore-1803 (based on 10.0.17134.766)
* mcr.microsoft.com/dotnet/framework/sdk:4.8-20190514-windowsservercore-ltsc2016 (based on 10.0.14393.2972)

**Comment by ablaylock** [comment]
I tested each of these and got the following results:

- mcr.microsoft.com/dotnet/framework/sdk:4.8-20190514-windowsservercore-ltsc2019 (based on 10.0.17763.504) **WORKS**
- mcr.microsoft.com/dotnet/framework/sdk:4.8-20190514-windowsservercore-1803 (based on 10.0.17134.766) **WORKS**
- mcr.microsoft.com/dotnet/framework/sdk:4.8-20190514-windowsservercore-ltsc2016 (based on 10.0.14393.2972) **DOESN'T WORK**

Attached is the vslogs.zip from the failed build based on ltsc2016. I re-read the host/container matrix and I *think* it should work. What about the ltsc2016 version causes it to fail?

[vslogs_ltsc2016.zip](https://github.com/MicrosoftDocs/visualstudio-docs/files/3204457/vslogs_ltsc2016.zip)

**Comment by heaths** [comment]
A download operation timed out. I will add a link to the documentation I provided above, but this no longer seems to be an issue with the documentation once that is in place. For general support, please visit https://developercommunity.visualstudio.com.

**Comment by ablaylock** [comment]
One remaining issue with the documentation is the documentation (I can open a new ticket if appropriate) is the Dockerfile doesn't correctly error out on this command:
```
RUN C:\TEMP\vs_buildtools.exe --quiet --wait --norestart --nocache `
    --installPath C:\BuildTools `
    --all `
    --remove Microsoft.VisualStudio.Component.Windows10SDK.10240 `
    --remove Microsoft.VisualStudio.Component.Windows10SDK.10586 `
    --remove Microsoft.VisualStudio.Component.Windows10SDK.14393 `
    --remove Microsoft.VisualStudio.Component.Windows81SDK `
 || IF "%ERRORLEVEL%"=="3010" EXIT 0
```

I have installations that will continue through that fail with error 255 as in the original issue description.

**Comment by heaths** [comment]
No, I thought that was working but on subsequent testing it is clearly not. I'm trying to fix that, but either my recent update to 19H1 and/or something in docker engine 18.09.2 isn't working when I try to reset the `SHELL` to anything else. I'm working with our virtualization team internally to determine the issue, but may be some time before I update the docs (don't want to update it to a _Dockerfile_ example I can't validate works).

**Comment by ablaylock** [comment]
Well for what it is worth I successfully built it on my system but I am pretty unsettled with the non-determinism of the same Dockerfile, it might not work the next time I try.

Here is my docker file (slightly modified to get the logs with collect.exe):

```
# escape=`

# Use the latest Windows Server Core image with .NET Framework 4.7.1.
FROM microsoft/dotnet-framework:4.7.1
RUN mkdir C:\BuildTools

# Restore the default Windows shell for correct batch processing.
SHELL ["cmd", "/S", "/C"]

ADD https://aka.ms/vscollect.exe C:\TEMP\collect.exe

# Download the Build Tools bootstrapper.
ADD https://aka.ms/vs/16/release/vs_buildtools.exe C:\TEMP\vs_buildtools.exe

# Install Build Tools excluding workloads and components with known issues.
RUN C:\TEMP\vs_buildtools.exe --quiet --wait --norestart --nocache `
    --installPath C:\BuildTools `
    --all `
    --remove Microsoft.VisualStudio.Component.Windows10SDK.10240 `
    --remove Microsoft.VisualStudio.Component.Windows10SDK.10586 `
    --remove Microsoft.VisualStudio.Component.Windows10SDK.14393 `
    --remove Microsoft.VisualStudio.Component.Windows81SDK `
 || IF "%ERRORLEVEL%"=="3010" EXIT 0
 
RUN C:\TEMP\collect.exe -zip:C:\vslogs.zip

# Start developer command prompt with any other commands specified.
ENTRYPOINT C:\BuildTools\Common7\Tools\VsDevCmd.bat &&

# Default to PowerShell if no other command specified.
CMD ["powershell.exe", "-NoLogo", "-ExecutionPolicy", "Bypass"]
```

**Comment by heaths** [comment]
As long as you have the right (supported) `FROM`, the only thing non-deterministic about this is:

1. If you use a lax tag - i.e. not specific to any particular build or KB support level of the OS - you might get a slightly newer OS version. As with any guest OS, it all depends on how specific you get. If you want to be really specific, use an image manifest hash (looks like @sha256:_hash_).
2. What version of Build Tools. Using that link, it's always the latest. We don't currently publish specific version links by design. If you want to keep a specific build, please see our docs about creating a network layout.
3. Transient errors. For example, in one run either you or another user had download issues about half way through the install.

Except for the third issue, the other two are within your control depending on how specific you want to get. The same is even true on linux container builds: most Dockerfiles I see don't specify specific versions of packages to get via `apt-get` for example, but they could.
